### DISCUSSION TITLE
petsc4py not installing due to too old of cython

### DISCUSSION THREAD

**Discussion by lindsayad** [firedrakeproject/firedrake#3591] [Install FAQ]
This seems to be the relevant excerpt from the install log. It looks like the cython version is deliberately pinned to this old version in the install script.
[firedrake-install.log](https://github.com/firedrakeproject/firedrake/files/15439996/firedrake-install.log)

```
Building wheels for collected packages: petsc4py
  Created temporary directory: /tmp/pip-wheel-p5emmi5v
  Destination directory: /tmp/pip-wheel-p5emmi5v
  Building wheel for petsc4py (pyproject.toml): started
  Running command Building wheel for petsc4py (pyproject.toml)
  running bdist_wheel
  running build
  running build_src
  removing Cython 0.29.36 from sys.modules
  fetching build requirement 'Cython >= 3.0.0'
  ********************************************************************************

    You need Cython >= 3.0.0 (you have version 0.29.36)

      $ /home/lindad/firedrake/bin/python -m pip install --upgrade cython

  ********************************************************************************
  error: unsatisfied build requirement 'Cython >= 3.0.0'
  error: subprocess-exited-with-error
  
  × Building wheel for petsc4py (pyproject.toml) did not run successfully.
  │ exit code: 1
  ╰─> See above for output.
```

**Comment by lindsayad** [comment]
I'm using my own PETSc. It looks like this is a well-known issue and should hopefully be solved soonish by https://github.com/firedrakeproject/firedrake/pull/3290

**Reply by connorjward** [reply]
Yes we are trying to land support for Cython 3, numpy 2 and Python 3.12 all together. We have everything working and there is just a performance regression to track down. Definitely should be quite soon.

**Reply by connorjward** [reply]
numpy 2.0 is due to be released June 16th so once that happens we will get everything merged, including Cython 3 support.

**Reply by lindsayad** [reply]
sounds great!

**Comment by EirikJaccheri** [comment]
Hi, 
Has the problem been fixed? Because i get a similar error on my Ubuntu 22.04.5 LTS computer, both when i pip install and when i tried to build from source. One thing i find curious is that it is reverting to an older version of numpy, even if i have numpy 2.1.2 installed. 

Collecting numpy<2
  Using cached numpy-1.26.4-cp310-cp310-manylinux_2_17_x86_64.manylinux2014_x86_64.whl (18.2 MB)

Full error:
`python3 -m pip install src/binding/petsc4py
Defaulting to user installation because normal site-packages is not writeable
Processing ./src/binding/petsc4py
  Installing build dependencies ... done
  Getting requirements to build wheel ... done
  Installing backend dependencies ... done
  Preparing metadata (pyproject.toml) ... done
Collecting numpy<2
  Using cached numpy-1.26.4-cp310-cp310-manylinux_2_17_x86_64.manylinux2014_x86_64.whl (18.2 MB)
Collecting petsc<3.23,>=3.22
  Using cached petsc-3.22.0-py3-none-linux_x86_64.whl
Building wheels for collected packages: petsc4py
  Building wheel for petsc4py (pyproject.toml) ... error
  error: subprocess-exited-with-error
  
  × Building wheel for petsc4py (pyproject.toml) did not run successfully.
  │ exit code: 1
  ╰─> [11 lines of output]
      running bdist_wheel
      running build
      running build_src
      ********************************************************************************
      
        You need Cython >= 3.0.0 (you have version 0.29.28)
      
          $ /usr/bin/python3 -m pip install --upgrade cython
      
      ********************************************************************************
      error: unsatisfied build requirement 'Cython >= 3.0.0'
      [end of output]
  
  note: This error originates from a subprocess, and is likely not a problem with pip.
  ERROR: Failed building wheel for petsc4py
Failed to build petsc4py
ERROR: Could not build wheels for petsc4py, which is required to install pyproject.toml-based projects`

**Reply by connorjward** [reply]
This is surprising. We now have full support for numpy 2 and cython 3. Are you using up-to-date install scripts?

**Reply by EirikJaccheri** [reply]
Hi,
Thanks for the quick response:-)

I followed this installation guide for installing petsc4py from source: https://petsc.org/release/petsc4py/install.html

I used the following command

$ python -m pip install src/binding/petsc4py

In the release branch of petsc in gitlab

git clone -b release https://gitlab.com/petsc/petsc.git petsc

Should this have the correct install scripts?

Eirik

**Reply by connorjward** [reply]
I would have thought so. However this is far from a standard way of installing PETSc for Firedrake:
* We install a number of external packages during PETSc configure like a direct solver and mesh partitioner.
* We install a fork of PETSc instead of the release version (we carry around one small patch, which we would love to get rid of).
* When installing petsc4py manually the command I usually run is: `python -m pip install --no-deps src/binding/petsc4py`. You need to pass `--no-deps` because otherwise it will try to install a pip version of PETSc which we don't want. You might also need to set `PETSC_DIR` and `PETSC_ARCH` environment variables.

May I ask why you aren't using the Firedrake install script?

**Reply by EirikJaccheri** [reply]
I only need petsc4py and not Firedrake.

I realize that this might not be the right forum to ask, but it was the only place i found someone having a similar issue...

I tried installing with --no-deops and got the same issue.

**Reply by connorjward** [reply]
Yes this would be better asked directly to PETSc: https://gitlab.com/petsc/petsc/-/issues.

As a last suggestion beforehand though have you simply tried upgrading Cython:
```
$ python -m pip install -U cython
```
